houdini/python2.7libs/conductor_houdini.py

```diff
@@ -0,0 +1,168 @@
+"""Handle callbacks and other actions from the node.
+
+This includes handling all actions from widgets, aux buttons
+as well as populating menus, initializing state and so on.
+
+"""
+
+import hou
+from hda import (
+    instances,
+    projects,
+    frame_spec,
+    render_source,
+    submit,
+    software,
+    stats,
+    notifications,
+    takes
+)
+
+from conductor import CONFIG
+
+if CONFIG.get("log_level") == "DEBUG":
+    reload(projects)
+    reload(instances)
+    reload(frame_spec)
+    reload(render_source)
+    reload(submit)
+    reload(software)
+    reload(stats)
+    reload(notifications)
+    reload(takes)
+
+
+def _update_node(node, **kw):
+    """Initialize or update.
+
+    We do this on creation/loading or manually.
+
+    """
+    projects.fetch(node)
+    instances.fetch_types(node)
+    frame_spec.validate_custom_range(node)
+    frame_spec.validate_scout_range(node)
+    render_source.update_input_node(node)
+    stats.update_estimates(node)
+    frame_spec.set_type(node)
+    notifications.validate_emails(node)
+    notifications.email_hook_changed(node)
+    takes.update_takes(node)
+    software.detect(node)
+
+
+MENUS = dict(
+    machine_type=instances.populate_menu,
+    project=projects.populate_menu
+)
+
+ACTIONS = dict(
```

Comment:
These are all names I have specified in the hda interface (which is saved as a binary file) so they will be there for sure. But I can add an except block for KeyError after hou.Error just in case.

Reply:
I did some tests and if a KeyError occurs, Houdini does not crash. It brings up a panel with a stack trace.  A KeyError would be very unexpected and could only happen if bad code was released, so I guess its best that Houdini handles it like it would other errors.